### DISCUSSION TITLE
can't retrieve last values for rke2

### DISCUSSION THREAD

**Discussion by sylvainOL** [renovatebot/renovate#15948] [Request Help]
Hello,
First thanks for the nice tool.
I managed to make it work in different ways but here I'm a bit stuck so I search for help!

I've got yaml files with the version of kubernetes I want to use:

```yaml
---
apiVersion: upgrade.cattle.io/v1
kind: Plan
metadata:
  labels:
    rke2-upgrade: agent
  name: agent-plan
  namespace: system-upgrade
spec:
  concurrency: 1
  cordon: true
  drain:
    force: true
  nodeSelector:
    matchExpressions:
    - key: rke2-upgrade
      operator: Exists
    - key: rke2-upgrade
      operator: NotIn
      values:
      - disabled
      - 'false'
    - key: node-role.kubernetes.io/control-plane
      operator: NotIn
      values:
      - 'true'
  prepare:
    args:
    - prepare
    - server-plan
    image: rancher/rke2-upgrade
  serviceAccountName: system-upgrade
  upgrade:
    image: rancher/rke2-upgrade
  version: v1.23.4+rke2r1
```

as you can see, I'm currently using `v1.23.4+rke2r1`
in github, the "last" versions (non rc ones) are `v1.23.6+rke2r2` and `v1.23.6+rke2r1`(see https://api.github.com/repos/rancher/rke2/releases)

in the renovate.json, I've tried several options in order to receive last release (`v1.23.6+rke2r2`) but I'm always retrieving the previous one (`v1.23.6+rke2r1`).

here's my last try:
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base",
    ":disableDependencyDashboard",
    ":semanticCommitTypeAll(fix)",
    ":disableRateLimiting"
  ],
  "separateMajorMinor": false,
  "regexManagers": [
    {
      "fileMatch": ["plan-.*\\.yaml$"],
      "datasourceTemplate": "github-releases",
      "matchStrings": [
        "version: (?<currentValue>.*?)\\n"
      ],
      "depNameTemplate": "rancher/rke2"
    }
  ],
  "packageRules": [
    {
      "matchPackageNames": ["rancher/rke2"],
      "versioning": "regex:^v1\\.(?<major>\\d+)\\.(?<minor>\\d+)(:?-?(?<prerelease>.*)?)\\+rke2r(?<patch>\\d+)\\n?$"
    },

    {
      "matchUpdateTypes": ["minor"],
      "extends": [
        ":semanticCommitTypeAll(feat)"
      ]
    },
    {
      "matchUpdateTypes": ["major"],
      "commitBody": "BREAKING CHANGE: new upstream major version",
      "extends": [
        ":semanticCommitTypeAll(feat)"
      ]
    },
    {
      "matchManagers": "gitlabci-include",
      "extends": [
        ":semanticCommitTypeAll(chore)"
      ]
    }
  ]
}
```

I've tried for versioning:

* `loose`
* `regex:^v(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(:?-?(?<prerelease>.*)?)\\+rke2r(?<build>\\d+)\\n?$`
* `regex:^v(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(:?-?(?<prerelease>.*)?)\\+rke2r(?<build>\\d+)$`
* `regex:^v(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(-?(?<prerelease>.*)?)\\+rke2r(?<build>\\d+)$`

I've also tried `regex:^v(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)(:?-?(?<prerelease>.*)?)\\+rke3r(?<build>\\d+)\\n?$` (I've changed rke**2** by rke**3**) and I've got an error.

Do you see why I can't have version `v1.23.6+rke2r2` as target?

I keep having something like:

```
DEBUG: Found regex package files (repository=repo/test-project)
DEBUG: Found 2 package file(s) (repository=repo/test-project)
 INFO: Dependency extraction complete (repository=repo/test-project)
       "baseBranch": "main",
       "stats": {
         "managers": {"regex": {"fileCount": 2, "depCount": 2}},
         "total": {"fileCount": 2, "depCount": 2}
       }
DEBUG: PackageFiles.add() - Package file saved for branch (repository=repo/test-project)
       "baseBranch": "main"
DEBUG: Package releases lookups complete (repository=repo/test-project)
       "baseBranch": "main"
DEBUG: branchifyUpgrades (repository=repo/test-project)
DEBUG: 2 flattened updates found: rancher/rke2, rancher/rke2 (repository=repo/test-project)
DEBUG: Returning 1 branch(es) (repository=repo/test-project)
DEBUG: Fetching changelog: https://github.com/rancher/rke2 (v1.23.4+rke2r1 -> v1.23.6+rke2r1) (repository=repo/test-project)
DEBUG: Fetching changelog: https://github.com/rancher/rke2 (v1.23.4+rke2r1 -> v1.23.6+rke2r1) (repository=repo/test-project)
DEBUG: config.repoIsOnboarded=true (repository=repo/test-project)
DEBUG: packageFiles with updates (repository=repo/test-project)
       "config": {
         "regex": [
           {
             "packageFile": "base/plan-agent.yaml",
             "deps": [
               {
                 "depName": "rancher/rke2",
                 "currentValue": "v1.23.4+rke2r1",
                 "datasource": "github-releases",
                 "replaceString": "version: v1.23.4+rke2r1\n",
                 "depIndex": 0,
                 "updates": [
                   {
                     "bucket": "latest",
                     "newVersion": "v1.23.6+rke2r1",
                     "newValue": "v1.23.6+rke2r1",
                     "releaseTimestamp": "2022-04-28T20:11:03.000Z",
                     "newMajor": 23,
                     "newMinor": 6,
                     "updateType": "minor",
                     "branchName": "renovate/rancher-rke2-23.x"
                   }
                 ],
                 "warnings": [],
                 "versioning": "regex:^v1\\.(?<major>\\d+)\\.(?<minor>\\d+)(:?-?(?<prerelease>.*)?)\\+rke2r(?<patch>\\d+)\\n?$",
                 "sourceUrl": "https://github.com/rancher/rke2",
                 "currentVersion": "v1.23.4+rke2r1",
                 "isSingleVersion": true,
                 "fixedVersion": "v1.23.4+rke2r1"
               }
             ],
             "matchStrings": ["version: (?<currentValue>.*?)\\n"],
             "depNameTemplate": "rancher/rke2",
             "datasourceTemplate": "github-releases"
           },
           {
             "packageFile": "base/plan-server.yaml",
             "deps": [
               {
                 "depName": "rancher/rke2",
                 "currentValue": "v1.23.4+rke2r1",
                 "datasource": "github-releases",
                 "replaceString": "version: v1.23.4+rke2r1\n",
                 "depIndex": 0,
                 "updates": [
                   {
                     "bucket": "latest",
                     "newVersion": "v1.23.6+rke2r1",
                     "newValue": "v1.23.6+rke2r1",
                     "releaseTimestamp": "2022-04-28T20:11:03.000Z",
                     "newMajor": 23,
                     "newMinor": 6,
                     "updateType": "minor",
                     "branchName": "renovate/rancher-rke2-23.x"
                   }
                 ],
                 "warnings": [],
                 "versioning": "regex:^v1\\.(?<major>\\d+)\\.(?<minor>\\d+)(:?-?(?<prerelease>.*)?)\\+rke2r(?<patch>\\d+)\\n?$",
                 "sourceUrl": "https://github.com/rancher/rke2",
                 "currentVersion": "v1.23.4+rke2r1",
                 "isSingleVersion": true,
                 "fixedVersion": "v1.23.4+rke2r1"
               }
             ],
             "matchStrings": ["version: (?<currentValue>.*?)\\n"],
             "depNameTemplate": "rancher/rke2",
             "datasourceTemplate": "github-releases"
           }
         ]
       }
```

I've also looked at trace logs but it's super verbose and I couldn't find any part of "parsing" version...

**Comment by rarkins** [comment]
My first guess is that prerelease tagging is interfering.

<img width="278" alt="image" src="https://user-images.githubusercontent.com/6311784/172988997-f0833223-b557-4f67-bbb0-0c726ef83015.png">

<img width="240" alt="image" src="https://user-images.githubusercontent.com/6311784/172989059-cd3275f7-e420-4558-81aa-9137154608cc.png">

<img width="244" alt="image" src="https://user-images.githubusercontent.com/6311784/172989645-639b78c5-6fb0-4815-8d19-d8306efc9852.png">

But both have prerelease=true, so maybe not. Do you have a way of telling from the logs if both versions are considered but they are simply sorted wrong?

@zharinov can you clarify how the `prerelease` tag in GitHub releases influences results?

**Comment by sylvainOL** [comment]
Hello @rarkins,

First thanks for taking time to see what I'm doing wrong.
I also thought about the `prerelease` tag  but both have them to `false` yes.
for the log, I tried to parse the 30Mb of logs trace is giving but wasn't able to find `1.23.6`before a log saying "I found `1.23.6+rke2r1` :/
on debug, we have nothing on the log for selection.
I also tried to look at code of both regex version and github releases but wasn't smart enough to see what can be wrong :(

**Reply by rarkins** [reply]
Can you set up a reproduction repo so that we can debug it at runtime? Then we can see if all expected versions are returned and how they are sorted

**Reply by sylvainOL** [reply]
yes of course!
I do that right away

**Comment by sylvainOL** [comment]
I've created https://github.com/sylvainOL/rke2-renovate
please bear in mind that my work is on a private gitlab instance. So I've shown `config.js` of test renovate I'm using and it's related gitlab ci file in the README

**Comment by lindhe** [comment]
Glad to see I'm not the only one struggling to get the latest version from RKE2. 😅 

I'm pretty surprised that Renovate does not pick up "v1.32.8+rke2r1" via the regular `semver` versioning. Arguably, `v` is not a valid prefix according to the [semver spec](https://semver.org/), but other than that it follows the spec to a tee. But even if I try to mask out the `v` via regex versioning or `extractVersion`, I can't get it to work.

I want to emphasize that `+rke2r1` is considered build metadata, but prerelease version, just so we're all in agreement on that.

Do we have any new ideas on why RKE2's semver does not work with Renovate?
